### PR TITLE
Refactor room permissions

### LINKED ISSUE #110: Allow users with permissions to view, edit, delete all rooms
**Is your feature request related to a problem? Please describe.**
Currently only the room owner can see all details of his room and edit the settings.
To provide better support for teachers and students the admin users should be able to see all rooms and edit the settings.

**Describe the solution you'd like**
Refactor room permission system. Also consider added a feature to allow moderators or a new room role to manage the room settings, members and files.

**Describe alternatives you've considered**

--

**Additional context**

--

### PR DESCRIPTION
**Pull request by SamuelWei**
# Fixes #110, #58

## Type (Highlight the corresponding type)
- Bugfix
- **Feature**
- Documentation
- **Refactoring** (e.g. Style updates, Test implementation, etc.)
- Other (please describe):

## Checklist
- [x] Code updated to current masters head
- [x] Passes CI checks
- [x] Is a part of an issue
- [x] Tests added for the bugfix or newly implemented feature, describe below why if not
- [x] Changelog is updated
- [x] Documentation of code and features exists

## Changes

- Added permission to manage all rooms
- Added readonly view of members, files and room settings for users with rooms.viewAll permission
- Added new room role "co-owner" that can change the room settings, files and members (e.g. two lecturer share the same room, teacher shares room with his tutors)
- Moved most of the permission logic to permission classes
- Fixed guests were unable to download files from a room with access code
- Changed the user details available through the user search

### CODE REVIEW

**Comment by dennis95stumm** [comment]
Please update to masters head!

**Comment by SamuelWei** [comment]
> Please update to masters head!

Done

File: resources/js/components/Room/MembersComponent.vue [hunk 1]
```diff
@@ -61,7 +64,7 @@
 
           <!-- action buttons -->
           <template v-slot:cell(actions)="data">
-            <b-button-group class="float-md-right">
+            <b-button-group class="float-md-right" v-if="currentUser.id !== data.item.id" >
```
Comment (dennis95stumm) [inline resources/js/components/Room/MembersComponent.vue:67]: If the user is on the room page and logs out it leads to an error!

**Review by dennis95stumm — CHANGES_REQUESTED** [review]

File: app/Room.php [hunk 2]
```diff
@@ -172,8 +184,8 @@ public function getRole($user)
             return RoomUserRole::GUEST;
         }
 
-        if ($this->owner->is($user)) {
-            return RoomUserRole::MODERATOR;
+        if ($this->owner->is($user) || $user->can('rooms.manage')) {
```
[the remark this answers is not in the document]
Comment (SamuelWei) [reply app/Room.php:187]: if RoomUserRole::OWNER is returned, that results in having the user join as moderator in the meeting, see getJoinUrl function. If the user has permission to manage a room, he could also add himself as moderator or co-owner and bypass any thing else that would be definied here.

[the remark this answers is not in the document]
Comment (dennis95stumm) [reply app/Room.php:187]: Yes but why a admin should join as a moderator?

File: resources/js/views/rooms/View.vue [hunk 3]
```diff
@@ -29,15 +29,16 @@
       <div class="row pt-7 pt-sm-9 mb-3" v-if="room.authenticated && isAuthenticated">
         <div class="col-lg-12">
           <!-- If membership is enabled, allow user to become member -->
+          <can method="becomeMember" :policy="room">
```
Comment (dennis95stumm) [inline resources/js/views/rooms/View.vue:32]: Here must the model be passed!

File: resources/js/views/rooms/View.vue [hunk 4]
```diff
@@ -3,7 +3,7 @@
     <template v-if="room !== null">
 
       <!-- Delete button and modal -->
-      <can method="delete" :policy="{ model_name: 'Room', isOwner: room.isOwner  }">
+      <can method="delete" :policy="room">
```
Comment (dennis95stumm) [inline resources/js/views/rooms/View.vue:6]: Here the model must be passed!

File: app/Http/Middleware/RoomAuthenticate.php [hunk 5]
```diff
@@ -27,7 +28,7 @@ public function handle($request, Closure $next, $allowAuthenticated = false)
         $room          = $request->route('room');
 
         // requested user is the owner or a member of the room or the room doesn't require access code
-        if ($room->owner->is(Auth::user()) or $room->members->contains(Auth::user()) or $room->accessCode == null) {
+        if ($room->accessCode == null || (Auth::user() && ($room->owner->is(Auth::user()) || $room->members->contains(Auth::user()) || Auth::user()->can('viewAll', Room::class)))) {
```
Comment (dennis95stumm) [inline app/Http/Middleware/RoomAuthenticate.php:31]: Is this extra `Auth::user()` check necessary?

File: resources/js/policies/RoomPolicy.js [hunk 6]
```diff
@@ -12,16 +12,62 @@ export default {
...
+  manageSettings (permissionService, model) {
+    return !permissionService.currentUser ? false : model.owner.id === permissionService.currentUser.id || model.isCoOwner || permissionService.currentUser.permissions.includes('rooms.manage');
+  },
+
+  /**
+   * Is user allowed to become member of this room
+   * @param permissionService
+   * @param model
+   */
+  becomeMember (permissionService, model) {
```
Comment (dennis95stumm) [inline resources/js/policies/RoomPolicy.js:69]: Please add tests!

File: app/Http/Resources/Room.php [hunk 7]
```diff
@@ -39,18 +39,21 @@ public function toArray($request)
...
                 'isMember'          => $this->resource->isMember(Auth::user()),
-                'isOwner'           => $this->owner->is(Auth::user()),
-                'isGuest'           => Auth::guest(),
-                'isModerator'       => $this->resource->isModeratorOrOwner(Auth::user()),
+                'isModerator'       => $this->resource->isModerator(Auth::user()),
+                'isCoOwner'         => $this->resource->isCoOwner(Auth::user()),
                 'canStart'          => Gate::inspect('start', $this->resource)->allowed(),
                 'running'           => $this->resource->runningMeeting() != null,
-                'accessCode'        => $this->when($this->resource->isModeratorOrOwner(Auth::user()), $this->accessCode),
+                'accessCode'        => $this->when(Gate::inspect('viewAccessCode', $this->resource)->allowed(), $this->accessCode),
```
Comment (dennis95stumm) [inline app/Http/Resources/Room.php:56]: Please add tests for this. When a user will or will not get this field.

File: app/Http/Middleware/RoomAuthenticate.php (hunk 5 above)
Comment (SamuelWei) [reply app/Http/Middleware/RoomAuthenticate.php:31]: Yes, if the accessCode is not null and the request comes from a guest Auth::user() would be null. Checking Auth::user()->can('viewAll', Room::class) would therefore result in an error.

Comment (dennis95stumm) [reply app/Http/Middleware/RoomAuthenticate.php:31]: But then the check is at a wrong place!

File: tests/Feature/api/v1/Room/RoomTest.php [hunk 8]
```diff
@@ -559,6 +722,28 @@ public function testStartOnlyModeratorsNoServer()
...
+        $room->members()->sync([]);
+
+        // Try with view all rooms permission
+        $this->user->roles()->attach($this->role);
+        $this->role->permissions()->attach($this->viewAllPermission);
+        $this->actingAs($this->user)->getJson(route('api.v1.rooms.start', ['room'=>$room]))
+            ->assertForbidden();
+        $this->role->permissions()->detach($this->viewAllPermission);
+
+        // Try with view all rooms permission
```
Comment (dennis95stumm) [inline tests/Feature/api/v1/Room/RoomTest.php:740]: Wrong comment. Should be something with manage permissions

File: tests/Feature/api/v1/Room/FileTest.php [hunk 9]
```diff
@@ -121,39 +147,59 @@ public function testViewFiles()
...
+            ->assertSuccessful()
+            ->assertJsonFragment(['filename'=>$this->file_valid->name]);
 
         // Testing owner
         $this->actingAs($this->room->owner)->getJson(route('api.v1.rooms.files.get', ['room'=>$this->room]))
             ->assertSuccessful()
             ->assertJsonFragment(['filename'=>$this->file_valid->name]);
 
+        // Remove membership roles and test with view all permission
+        $this->room->members()->sync([]);
```
Comment (dennis95stumm) [inline tests/Feature/api/v1/Room/FileTest.php:184]: What about the rooms.manage permission. Add a test for it!

File: app/Policies/RoomPolicy.php [hunk 10]
```diff
@@ -163,6 +198,14 @@ public function manageFiles(?User $user, Room $room)
      */
     public function downloadFile(?User $user, Room $room, RoomFile $roomFile)
     {
-        return $room->owner->is($user) || $roomFile->download === true;
+        if ($roomFile->download === true) {
```
Comment (dennis95stumm) [inline app/Policies/RoomPolicy.php:201]: Should guests have the permission to download files?

File: resources/js/components/Room/MembersComponent.vue [hunk 11]
```diff
@@ -439,7 +474,7 @@ export default {
 
     // member tables headings
     tablefields () {
-      return [
+      var fields = [
```
Comment (dennis95stumm) [inline resources/js/components/Room/MembersComponent.vue:477]: Please use const!

**Review by dennis95stumm — CHANGES_REQUESTED** [review]

File: app/Policies/RoomPolicy.php (hunk 10 above)
Comment (SamuelWei) [reply app/Policies/RoomPolicy.php:201]: Yes, see "Changes" section of this PR description

File: tests/Feature/api/v1/Room/FileTest.php (hunk 9 above)
Comment (SamuelWei) [reply tests/Feature/api/v1/Room/FileTest.php:184]: As the rooms.viewAll permission is inherited by the rooms.manage permission, I don't see a reason to check it again. I did not add these redundant tests in a few places. To check that the permission inheritance is working, there is the 'testPermissionInheritances' in tests/Feature/api/v1/Room/RoomTest.php

**Review by dennis95stumm — APPROVED** [review]

**Pull request merged**